### PR TITLE
Send the JSON state representation to Cloud backend (when available)

### PR DESCRIPTION
**Pull request by megan07**

### CODE REVIEW

File: internal/command/state_mv.go
Comment (alisdair) [inline internal/command/state_mv.go:401]: We might want to adjust the control flow here; `loadConfig(path)` shouldn't be called if `os.Getwd` errors, and we don't want to print the same warning three times. Same for other instances of this pattern.

File: internal/cloud/state.go
Comment (megan07) [inline internal/cloud/state.go:213]: @brandonc I added this in to fix the crash that I was getting on the cloud test. It seems as though `JSONStateOutputs` needs to be populated here, so I just copied what was in [backend_state.go](https://github.com/hashicorp/terraform/blob/5da30c2b65265c9c7ac7580a1295e87715ebb568/internal/cloud/backend_state.go#L66-L79) to do that.

* Is this the correct thing to do and is this the right spot? Wasn't sure if I was supposed to be putting this logic in `PersistState` instead - maybe it doesn't matter?
* Also, do I need to add `JSONState` in `backend_state.go`?

I think this also somehow/some way ties into why I needed to comment out [this test](https://github.com/hashicorp/terraform/pull/31698/files#diff-028e6f10c867390b9523c208ae67c17a730615e3a1e54600cae030c810ecedabR23-R64), but wasn't sure how to fix that.

Comment (brandonc) [reply internal/cloud/state.go:213]: Hmmm I think backend_state.go and state.go are doing the same thing and only one needs to exist. I think it's probably best to delete one altogether and move your changes to the other.

backend_state.go was added in the original PR that was not merged, but after that I added the simpler state.go which was just designed to wrap state persistence in order to save the json outputs while backend_state was designed to be a complete state interface (without an underlying delegate)

I'm not even sure which one is better, but state.go is simpler and already merged. If it can work for you, I would use it.

File: go.mod
Comment (brandonc) [inline go.mod:43]: Pre-merge TODO: release go-tfe and bump this ref to the released version

File: internal/cloud/state.go
Comment (brandonc) [inline internal/cloud/state.go:36]: ✂️ This can be deleted

File: internal/cloud/state.go
Comment (brandonc) [inline internal/cloud/state.go:145]: We should godoc each public method, making a comment like "PersistState uploads a snapshot the latest state as a StateVersion to Terraform Cloud"

File: internal/cloud/state_test.go
Comment (brandonc) [inline internal/cloud/state_test.go:107]: This part of the payload should be only the outputs object.

File: internal/cloud/testing.go
Comment (brandonc) [inline internal/cloud/testing.go:448]: This should be moved to state_test.go and probably not called as a helper but exist as a top level test.

File: internal/command/helper.go
Comment (brandonc) [inline internal/command/helper.go:10]:
```suggestion
external JSON variant of the state file. This will not have any adverse effects on Terraform's ability
```

File: internal/command/helper.go
Comment (brandonc) [inline internal/command/helper.go:13]: IMO we should end this warning at `on the next successful "terraform apply".` because the rest sounds unnecessarily scary. I would appreciate a second opinion!

File: internal/cloud/testing.go
Comment (brandonc) [reply internal/cloud/testing.go:448]: Not sure this test is called at all right now

File: internal/command/import.go
Comment (jbardin) [inline internal/command/import.go:256]: I think these are going to need to error out. The json marshaling code will panic if a nil schema is pass in.

Comment (jbardin) [reply internal/command/import.go:256]: Actually I take that back, digging deeper it looks like there were some nil checks added because that's happened before, and it _should_ error out instead, though I don't think purposely sending `nil` is something that was will tested.

File: internal/command/meta.go
Comment (jbardin) [inline internal/command/meta.go:824]: This looks like it can return `nil, nil`, which is a bit unusual. Is that going to be a problem anywhere?

File: internal/command/meta_backend_migrate.go
Comment (jbardin) [inline internal/command/meta_backend_migrate.go:443]: In case it matters for future ideas, the backend is currently handled before providers are installed during `init`, so requiring schemas here could lead to a catch-22 where it requires some manual intervention to proceed far enough for provider installation.

Comment (brandonc) [reply internal/command/meta_backend_migrate.go:443]: Perhaps the "hassle" wording you used could evolve into this observation. Useful insight

File: internal/command/helper.go
Comment (apparentlymart) [reply internal/command/helper.go:13]: Hmm indeed this message is tricky because it's introducing what are probably a bunch of entirely new concepts from the user's perspective, and doing so at an "anxiety point" where we're telling the user about some potential negative consequences they need to make a decision about. :confounded: 

I don't think there's a perfect way to do this because it's _never_ ideal to introduce entirely new concepts in an error/warning message, but I think I'd suggest reordering the information here a bit and splitting it over multiple paragraphs so it's hopefully a little easier to skim:

```
Warning: Failed to update data for external integrations

Terraform was unable to generate a description of the updated
state for use with external integrations in Terraform Cloud.
Any integrations configured for this workspace which depend on
information from the state may not work correctly when using the
result of this action.

This problem occurs when Terraform cannot read the schema for
one or more of the providers used in the state. The next successful
apply will correct the problem by re-generating the JSON description
of the state:
    terraform apply
```

(It's hard to get word-wrapping write in the GitHub comment box; I probably didn't wrap the lines above correctly for presentation in a monospace font.)

My goal with this edit was to put the "what happend and why it matters" part of the message first, and then the "why it happened and what to do about it" part second.

As a secondary goal I tried to omit the implementation detail about there being two different state serializations and instead focus on the fact that this missing information is intended for external integrations. I don't expect the typical reader of this message to know that external integrations use our JSON state format or even that there _is_ a JSON state format for integrations, and so mentioning it seems likely to leave the reader unsure of what the consequences of this situation are.

File: internal/command/meta.go
Comment (apparentlymart) [reply internal/command/meta.go:824]: From reading the various callsites of this function, I think I share the concern of a "nil with no errors" result being unexpected per typical Go idiom, which may cause confusion for future maintainers.

As a "quick fix" without doing any more significant immediate refactoring, I would suggest renaming this method to `MaybeGetSchemas` and mention in its documentation comment that it can potentially return `nil` without errors to indicate that there isn't enough information to return the schemas and that the caller must handle the lack of schema information itself accordingly. My hope is that an unusual method name with `Maybe` at the start of it will stand out enough to a future reader to cause them to investigate what the "maybe" implies, rather than just assuming this is a normal-ish function which will return errors if it isn't able to do its work.

File: internal/command/untaint.go
Comment (apparentlymart) [inline internal/command/untaint.go:171]: It feels kinda unfortunate that this idea of "cloud mode" is infecting so many codepaths now. I don't really have a better idea to suggest, though. :confounded: Hopefully we can think about a more elaborate refactoring of this later, but this surgical approach is probably best for now.

File: internal/command/untaint.go
Comment (apparentlymart) [inline internal/command/untaint.go:174]: We don't really typically use `c.Ui.Warn` for new code because it generates yellow text that can be hard to read in some terminals.

In another comment elsewhere I suggested renaming `c.GetSchemas` to `c.MaybeGetSchemas` to try to be explicit that there's "something odd" about this function -- that it can potentially fail but still return successfully in some sense. In conjunction with that, I wonder about making that function itself return this warning in the `tfdiags.Diagnostics` it returns, so that we could then render it with our usual diagnostic formatting codepath. (`c.showDiagnostics(diags)`, similar to some of the examples above)

**Review by apparentlymart — APPROVED** [review]
Due to the timing and broad scope of this change, @megan07 and I have agreed in a backchannel to move forward with this change in its current state and then potentially address some of the feedback items in subsequent separate PRs that we expect to be more localized, once some background discussions have played out and it seems clearer which path to take.

**Pull request merged**
